File: examples/typescript/facilitator/advanced/bazaar.ts
/**
 * Facilitator with Discovery Extension Example
 *
 * Demonstrates how to create a facilitator with bazaar discovery extension that
 * catalogs discovered x402 resources.
 */

import { base58 } from "@scure/base";
import { createKeyPairSignerFromBytes } from "@solana/kit";
import { x402Facilitator } from "@x402/core/facilitator";
import {
  PaymentPayload,
  PaymentRequirements,
  SettleResponse,
  VerifyResponse,
} from "@x402/core/types";
import { toFacilitatorEvmSigner } from "@x402/evm";
import { ExactEvmScheme } from "@x402/evm/exact/facilitator";
import { toFacilitatorSvmSigner } from "@x402/svm";
import { ExactSvmScheme } from "@x402/svm/exact/facilitator";
import { extractDiscoveryInfo, DiscoveryInfo } from "@x402/extensions/bazaar";
import dotenv from "dotenv";
import express from "express";
import { createWalletClient, http, publicActions } from "viem";
import { privateKeyToAccount } from "viem/accounts";
import { baseSepolia } from "viem/chains";

dotenv.config();

// Configuration
const PORT = process.env.PORT || "4022";

// Configuration - optional per network
const evmPrivateKey = process.env.EVM_PRIVATE_KEY as `0x${string}` | undefined;
const svmPrivateKey = process.env.SVM_PRIVATE_KEY as string | undefined;

// Validate at least one private key is provided
if (!evmPrivateKey && !svmPrivateKey) {
  console.error(
    "❌ At least one of EVM_PRIVATE_KEY or SVM_PRIVATE_KEY is required",
  );
  process.exit(1);
}

// Network configuration
const EVM_NETWORK = "eip155:84532"; // Base Sepolia
const SVM_NETWORK = "solana:EtWTRABZaYq6iMfeYKouRu166VU2xqa1"; // Solana Devnet

// DiscoveredResource represents a discovered x402 resource for the bazaar catalog
interface DiscoveredResource {
  resource: string;
  description?: string;
  mimeType?: string;
  type: string;
  x402Version: number;
  accepts: PaymentRequirements[];
  discoveryInfo?: DiscoveryInfo;
  lastUpdated: string;
}

// BazaarCatalog stores discovered resources
/**
 * Catalog of discovered resources from bazaar discovery extension.
 */
class BazaarCatalog {
  private resources: Map<string, DiscoveredResource> = new Map();

  /**
   * Adds a discovered resource to the catalog.
   *
   * @param res - The discovered resource to add
   */
  add(res: DiscoveredResource): void {
    this.resources.set(res.resource, res);
  }

  /**
   * Returns all discovered resources in the catalog.
   *
   * @returns Array of all discovered resources
   */
  getAll(): DiscoveredResource[] {
    return Array.from(this.resources.values());
  }
}

const bazaarCatalog = new BazaarCatalog();

// Initialize the x402 Facilitator with discovery hooks
const facilitator = new x402Facilitator()
  .onBeforeVerify(async (context) => {
    console.log("Before verify", context);
  })
  .onAfterVerify(async (context) => {
    console.log("✅ Payment verified");

    // Extract discovered resource from payment for bazaar catalog
    try {
      const discovered = extractDiscoveryInfo(
        context.paymentPayload,
        context.requirements,
        true, // validate
      );

      if (discovered) {
        console.log(`   📝 Discovered resource: ${discovered.resourceUrl}`);
        console.log(`   📝 Description: ${discovered.description}`);
        console.log(`   📝 MimeType: ${discovered.mimeType}`);
        if ("method" in discovered && discovered.method !== undefined) {
          console.log(`   📝 Method: ${discovered.method}`);
        } else if ("toolName" in discovered) {
          console.log(`   📝 Tool: ${discovered.toolName}`);
        }
        console.log(`   📝 X402Version: ${discovered.x402Version}`);

        bazaarCatalog.add({
          resource: discovered.resourceUrl,
          description: discovered.description,
          mimeType: discovered.mimeType,
          type: "http",
          x402Version: discovered.x402Version,
          accepts: [context.requirements],
          discoveryInfo: discovered.discoveryInfo,
          lastUpdated: new Date().toISOString(),
        });
        console.log("   ✅ Added to bazaar catalog");
      }
    } catch (err) {
      console.log(`   ⚠️  Failed to extract discovery info: ${err}`);
    }
  })
  .onVerifyFailure(async (context) => {
    console.log("Verify failure", context);
  })
  .onBeforeSettle(async (context) => {
    console.log("Before settle", context);
  })
  .onAfterSettle(async (context) => {
    console.log(`🎉 Payment settled: ${context.result.transaction}`);
  })
  .onSettleFailure(async (context) => {
    console.log("Settle failure", context);
  });

// Register EVM scheme if private key is provided
if (evmPrivateKey) {
  const evmAccount = privateKeyToAccount(evmPrivateKey);
  console.info(`EVM Facilitator account: ${evmAccount.address}`);

  // Create a Viem client with both wallet and public capabilities
  const viemClient = createWalletClient({
    account: evmAccount,
    chain: baseSepolia,
    transport: http(),
  }).extend(publicActions);

  const evmSigner = toFacilitatorEvmSigner({
    getCode: (args: { address: `0x${string}` }) => viemClient.getCode(args),
    address: evmAccount.address,
    readContract: (args: {
      address: `0x${string}`;
      abi: readonly unknown[];
      functionName: string;
      args?: readonly unknown[];
    }) =>
      viemClient.readContract({
        ...args,
        args: args.args || [],
      }),
    verifyTypedData: (args: {
      address: `0x${string}`;
      domain: Record<string, unknown>;
      types: Record<string, unknown>;
      primaryType: string;
      message: Record<string, unknown>;
      signature: `0x${string}`;
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    }) => viemClient.verifyTypedData(args as any),
    writeContract: (args: {
      address: `0x${string}`;
      abi: readonly unknown[];
      functionName: string;
      args: readonly unknown[];
    }) =>
      viemClient.writeContract({
        ...args,
        args: args.args || [],
      }),
    sendTransaction: (args: { to: `0x${string}`; data: `0x${string}` }) =>
      viemClient.sendTransaction(args),
    waitForTransactionReceipt: (args: { hash: `0x${string}` }) =>
      viemClient.waitForTransactionReceipt(args),
  });

  facilitator.register(
    EVM_NETWORK,
    new ExactEvmScheme(evmSigner, { deployERC4337WithEIP6492: true }),
  );
}

// Register SVM scheme if private key is provided
if (svmPrivateKey) {
  const svmAccount = await createKeyPairSignerFromBytes(
    base58.decode(svmPrivateKey),
  );
  console.info(`SVM Facilitator account: ${svmAccount.address}`);

  const svmSigner = toFacilitatorSvmSigner(svmAccount);

  facilitator.register(SVM_NETWORK, new ExactSvmScheme(svmSigner));
}

// Initialize Express app
const app = express();
app.use(express.json());

/**
 * POST /verify
 * Verify a payment against requirements
 *
 * Note: Payment tracking and bazaar discovery are handled by lifecycle hooks
 */
app.post("/verify", async (req, res) => {
  try {
    const { paymentPayload, paymentRequirements } = req.body as {
      paymentPayload: PaymentPayload;
      paymentRequirements: PaymentRequirements;
    };

    if (!paymentPayload || !paymentRequirements) {
      return res.status(400).json({
        error: "Missing paymentPayload or paymentRequirements",
      });
    }

    // Hooks will automatically:
    // - Track verified payment (onAfterVerify)
    // - Extract and catalog discovery info (onAfterVerify)
    const response: VerifyResponse = await facilitator.verify(
      paymentPayload,
      paymentRequirements,
    );

    res.json(response);
  } catch (error) {
    console.error("Verify error:", error);
    res.status(500).json({
      error: error instanceof Error ? error.message : "Unknown error",
    });
  }
});

/**
 * POST /settle
 * Settle a payment on-chain
 */
app.post("/settle", async (req, res) => {
  try {
    const { paymentPayload, paymentRequirements } = req.body;

    if (!paymentPayload || !paymentRequirements) {
      return res.status(400).json({
        error: "Missing paymentPayload or paymentRequirements",
      });
    }

    const response: SettleResponse = await facilitator.settle(
      paymentPayload as PaymentPayload,
      paymentRequirements as PaymentRequirements,
    );

    res.json(response);
  } catch (error) {
    console.error("Settle error:", error);

    // Check if this was an abort from hook
    if (
      error instanceof Error &&
      error.message.includes("Settlement aborted:")
    ) {
      return res.json({
        success: false,
        errorReason: error.message.replace("Settlement aborted: ", ""),
        network: req.body?.paymentPayload?.network || "unknown",
      } as SettleResponse);
    }

    res.status(500).json({
      error: error instanceof Error ? error.message : "Unknown error",
    });
  }
});

/**
 * GET /supported
 * Get supported payment kinds and extensions
 */
app.get("/supported", async (req, res) => {
  try {
    const response = facilitator.getSupported();
    res.json(response);
  } catch (error) {
    console.error("Supported error:", error);
    res.status(500).json({
      error: error instanceof Error ? error.message : "Unknown error",
    });
  }
});

/**
 * GET /discovery/resources
 * List all discovered resources from bazaar
 */
app.get("/discovery/resources", async (req, res) => {
  try {
    const resources = bazaarCatalog.getAll();
    res.json({
      x402Version: 2,
      items: resources,
      pagination: {
        limit: 100,
        offset: 0,
        total: resources.length,
      },
    });
  } catch (error) {
    console.error("Discovery error:", error);
    res.status(500).json({
      error: error instanceof Error ? error.message : "Unknown error",
    });
  }
});

/**
 * GET /health
 * Health check endpoint
 */
app.get("/health", (req, res) => {
  res.json({ status: "ok" });
});

// Start the server
app.listen(parseInt(PORT), () => {
  console.log(`🚀 Discovery Facilitator listening on http://localhost:${PORT}`);
  console.log(
    `   Supported networks: ${facilitator
      .getSupported()
      .kinds.map((k) => k.network)
      .join(", ")}`,
  );
  console.log(`   Discovery endpoint: GET /discovery/resources`);
  console.log();
});

File: examples/python/facilitator/advanced/bazaar.py
"""Facilitator with Discovery Extension Example.

Demonstrates how to create a facilitator with bazaar discovery extension that
catalogs discovered x402 resources.
"""

import os
import sys
import base64
import json
from datetime import UTC, datetime
from typing import Any

from dotenv import load_dotenv
from fastapi import FastAPI, HTTPException, Response
from pydantic import BaseModel
from solders.keypair import Keypair

from x402 import x402Facilitator
from x402.extensions.bazaar import (
    DiscoveryResource,
    extract_discovery_info,
)
from x402.mechanisms.evm import FacilitatorWeb3Signer
from x402.mechanisms.evm.exact.facilitator import ExactEvmScheme, ExactEvmSchemeConfig
from x402.mechanisms.svm import FacilitatorKeypairSigner
from x402.mechanisms.svm.exact.facilitator import ExactSvmScheme

# Load environment variables
load_dotenv()

# Configuration
PORT = int(os.environ.get("PORT", "4022"))

# Configuration - optional per network
evm_private_key = os.environ.get("EVM_PRIVATE_KEY")
svm_private_key = os.environ.get("SVM_PRIVATE_KEY")

# Validate at least one private key is provided
if not evm_private_key and not svm_private_key:
    print("❌ At least one of EVM_PRIVATE_KEY or SVM_PRIVATE_KEY is required")
    sys.exit(1)

# Network configuration
EVM_NETWORK = "eip155:84532"  # Base Sepolia
SVM_NETWORK = "solana:EtWTRABZaYq6iMfeYKouRu166VU2xqa1"  # Solana Devnet


# BazaarCatalog stores discovered resources
class BazaarCatalog:
    """Catalog for storing discovered x402 resources."""

    def __init__(self) -> None:
        self.resources: dict[str, DiscoveryResource] = {}

    def add(self, resource: DiscoveryResource) -> None:
        """Add a resource to the catalog."""
        self.resources[resource.resource] = resource

    def get_all(self) -> list[DiscoveryResource]:
        """Get all resources in the catalog."""
        return list(self.resources.values())

    def search(
        self,
        query: str,
        resource_type: str | None = None,
        limit: int | None = None,
    ) -> list[DiscoveryResource]:
        """Search resources using case-insensitive keyword matching.

        Matches against resource URL, type, description, service metadata, and extensions.

        Args:
            query: The search query string.
            resource_type: Optional filter by resource type.
            limit: Optional advisory maximum results.

        Returns:
            Matching resources.
        """
        needle = query.lower()
        results = []
        for r in self.resources.values():
            haystack = " ".join(
                [
                    r.resource,
                    r.type,
                    r.description or "",
                    r.service_name or "",
                    *(r.tags or []),
                    *[str(v) for v in (r.extensions or {}).values()],
                ]
            ).lower()
            if needle in haystack:
                results.append(r)

        if resource_type:
            results = [r for r in results if r.type == resource_type]

        return results[:limit] if limit is not None else results


bazaar_catalog = BazaarCatalog()

EXTENSION_RESPONSES_HEADER = "EXTENSION-RESPONSES"


def _set_extension_responses_header(response: Response) -> None:
    """Attach an example bazaar extension response header for client readback."""
    extension_responses = {"bazaar": {"status": "success"}}
    encoded = base64.b64encode(json.dumps(extension_responses).encode("utf-8")).decode("ascii")
    response.headers[EXTENSION_RESPONSES_HEADER] = encoded

# Initialize signers based on available keys
evm_signer = None
svm_signer = None

if evm_private_key:
    evm_signer = FacilitatorWeb3Signer(
        private_key=evm_private_key,
        rpc_url=os.environ.get("EVM_RPC_URL", "https://sepolia.base.org"),
    )
    print(f"EVM Facilitator account: {evm_signer.get_addresses()[0]}")

if svm_private_key:
    svm_keypair = Keypair.from_base58_string(svm_private_key)
    svm_signer = FacilitatorKeypairSigner(svm_keypair)
    print(f"SVM Facilitator account: {svm_signer.get_addresses()[0]}")


def _handle_after_verify(ctx: Any) -> None:
    """Handle after verify hook - extract discovery info and catalog."""
    print("✅ Payment verified")

    # Extract discovered resource from payment for bazaar catalog
    try:
        discovered = extract_discovery_info(
            ctx.payment_payload,
            ctx.requirements,
            validate=True,
        )

        if discovered:
            print(f"   📝 Discovered resource: {discovered.resource_url}")
            print(f"   📝 Method: {discovered.method}")
            print(f"   📝 X402Version: {discovered.x402_version}")
            if discovered.service_name is not None:
                print(f"   📝 Service: {discovered.service_name}")
            if discovered.tags is not None:
                print(f"   📝 Tags: {', '.join(discovered.tags)}")

            bazaar_catalog.add(
                DiscoveryResource(
                    resource=discovered.resource_url,
                    type=discovered.discovery_info.input.type,
                    x402_version=discovered.x402_version,
                    accepts=[
                        ctx.requirements.model_dump(by_alias=True)
                        if hasattr(ctx.requirements, "model_dump")
                        else ctx.requirements
                    ],
                    last_updated=datetime.now(UTC).isoformat().replace("+00:00", "Z"),
                    description=discovered.description,
                    mime_type=discovered.mime_type,
                    service_name=discovered.service_name,
                    tags=discovered.tags,
                    icon_url=discovered.icon_url,
                    extensions=discovered.extensions,
                )
            )
            print("   ✅ Added to bazaar catalog")
    except Exception as err:
        print(f"   ⚠️  Failed to extract discovery info: {err}")


# Initialize the x402 Facilitator with discovery hooks
facilitator = (
    x402Facilitator()
    .on_before_verify(lambda ctx: print("Before verify", ctx))
    .on_after_verify(lambda ctx: _handle_after_verify(ctx))
    .on_verify_failure(lambda ctx: print("Verify failure", ctx))
    .on_before_settle(lambda ctx: print("Before settle", ctx))
    .on_after_settle(lambda ctx: print(f"🎉 Payment settled: {ctx.result.transaction}"))
    .on_settle_failure(lambda ctx: print("Settle failure", ctx))
)

# Register schemes based on available signers
if evm_signer:
    config = ExactEvmSchemeConfig(
        # Add trusted ERC-6492 factory addresses here (e.g. your chosen ERC-4337 smart wallet factory).
        # A non-empty list enables smart wallet deployment; an empty list denies all factory calls.
        eip6492_allowed_factories=[],
    )
    facilitator.register([EVM_NETWORK], ExactEvmScheme(evm_signer, config))

if svm_signer:
    facilitator.register([SVM_NETWORK], ExactSvmScheme(svm_signer))


# Pydantic models for request/response
class VerifyRequest(BaseModel):
    """Verify endpoint request body."""

    paymentPayload: dict
    paymentRequirements: dict


class SettleRequest(BaseModel):
    """Settle endpoint request body."""

    paymentPayload: dict
    paymentRequirements: dict


# Initialize FastAPI app
app = FastAPI(
    title="Discovery Facilitator",
    description="Verifies and settles x402 payments with bazaar discovery",
    version="2.0.0",
)


@app.post("/verify")
async def verify(request: VerifyRequest, http_response: Response):
    """Verify a payment against requirements.

    Note: Payment tracking and bazaar discovery are handled by lifecycle hooks.

    Args:
        request: Payment payload and requirements to verify.

    Returns:
        VerifyResponse with isValid and payer (if valid) or invalidReason.
    """
    try:
        from x402.schemas import PaymentRequirements, parse_payment_payload

        # Parse payload (auto-detects V1/V2) and requirements
        payload = parse_payment_payload(request.paymentPayload)
        requirements = PaymentRequirements.model_validate(request.paymentRequirements)

        # Hooks will automatically:
        # - Track verified payment (on_after_verify)
        # - Extract and catalog discovery info (on_after_verify)
        verify_result = await facilitator.verify(payload, requirements)

        _set_extension_responses_header(http_response)
        return verify_result.model_dump(by_alias=True, exclude_none=True)
    except Exception as e:
        print(f"Verify error: {e}")
        raise HTTPException(status_code=500, detail=str(e)) from e


@app.post("/settle")
async def settle(request: SettleRequest, http_response: Response):
    """Settle a payment on-chain.

    Args:
        request: Payment payload and requirements to settle.

    Returns:
        SettleResponse with success, transaction, network, and payer.
    """
    try:
        from x402.schemas import PaymentRequirements, parse_payment_payload

        # Parse payload (auto-detects V1/V2) and requirements
        payload = parse_payment_payload(request.paymentPayload)
        requirements = PaymentRequirements.model_validate(request.paymentRequirements)

        settle_result = await facilitator.settle(payload, requirements)

        _set_extension_responses_header(http_response)
        return settle_result.model_dump(by_alias=True, exclude_none=True)
    except Exception as e:
        print(f"Settle error: {e}")

        # Check if this was an abort from hook
        if "aborted" in str(e).lower() or "Settlement aborted" in str(e):
            from x402.schemas import SettleResponse

            abort = SettleResponse(
                success=False,
                error_reason=str(e).replace("Settlement aborted: ", ""),
                network=request.paymentPayload.get("accepted", {}).get("network", "unknown"),
                transaction="",
            )
            return abort.model_dump(by_alias=True, exclude_none=True)

        raise HTTPException(status_code=500, detail=str(e)) from e


@app.get("/supported")
async def supported():
    """Get supported payment kinds and extensions.

    Returns:
        SupportedResponse with kinds, extensions, and signers.
    """
    try:
        response = facilitator.get_supported()

        return {
            "kinds": [k.model_dump(by_alias=True, exclude_none=True) for k in response.kinds],
            "extensions": response.extensions,
            "signers": response.signers,
        }
    except Exception as e:
        print(f"Supported error: {e}")
        raise HTTPException(status_code=500, detail=str(e)) from e


@app.get("/discovery/resources")
async def discovery_resources():
    """List all discovered resources from bazaar.

    Returns:
        Discovery response with x402Version, items, and pagination.
    """
    try:
        resources = bazaar_catalog.get_all()
        return {
            "x402Version": 2,
            "items": [r.to_dict() for r in resources],
            "pagination": {
                "limit": 100,
                "offset": 0,
                "total": len(resources),
            },
        }
    except Exception as e:
        print(f"Discovery error: {e}")
        raise HTTPException(status_code=500, detail=str(e)) from e


@app.get("/discovery/search")
async def discovery_search(query: str, type: str | None = None, limit: int | None = None):
    """Search discovered resources using keyword matching.

    Args:
        query: The search query string.
        type: Optional filter by resource type.
        limit: Optional advisory maximum number of results.

    Returns:
        Search response with x402Version, items, and optional pagination hints.
    """
    try:
        results = bazaar_catalog.search(query, type, limit)
        return {
            "x402Version": 2,
            "resources": [r.to_dict() for r in results],
            "partialResults": False,
            "pagination": None,
        }
    except Exception as e:
        print(f"Discovery search error: {e}")
        raise HTTPException(status_code=500, detail=str(e)) from e


@app.get("/health")
async def health():
    """Health check endpoint."""
    return {"status": "ok"}


if __name__ == "__main__":
    import uvicorn

    supported_networks = [k.network for k in facilitator.get_supported().kinds]
    print(f"🚀 Discovery Facilitator listening on http://0.0.0.0:{PORT}")
    print(f"   Supported networks: {', '.join(supported_networks)}")
    print("   Discovery endpoint: GET /discovery/resources")
    print()
    uvicorn.run(app, host="0.0.0.0", port=PORT)
